cadcutils/cadcutils/net/auth.py

```diff
@@ -69,25 +69,103 @@
+        """
+            The subject is anonymous if neither of this arguments is set
+        :param username: user name
+        :param certificate: name of the X509 certificate file
+        :param netrc: use information from .netrc. Value can be True (use default $HOME/.netrc)
+                    or the name of the netrc file to use.
+        """
+        self.logger = logging.getLogger('Subject')
+        self.username = username
+        self.certificate = certificate
+        self.netrc = netrc
+        self.hosts_auth = {}
+        if certificate is not None:
+            assert certificate is not '' and os.path.isfile(certificate)
+        else:
+            if netrc:
+                hosts = netrclib.netrc(netrc if netrc is not True else None).hosts
+                for host_name in hosts:
+                    self.hosts_auth[host_name] = (hosts[host_name][0], hosts[host_name][2])
+        self.username = username
+        self.anon = (self.certificate is None) and (self.netrc is False) \
+            and (username is None)
+
+    @staticmethod
+    def get_subject(args):
+        """
+        Instantiates a subject based on attributes of a command line. It works with the base parser in cadcutils.
+        and uses the following command line arguments:
+            args.user: username
+            args.cert: x509 certificate location
+            args.n: use netrc files for authentication info
+            args.netrc_file: use this netrc file for authentication info
+        :param args: argparse command line arguments
+        :return: corresponding subject
+        """
+        return Subject(username=args.user, certificate=args.cert,
+                       netrc=(args.netrc_file if args.netrc_file is not None else args.n))
+
+    def get_auth(self, realm):
+        """
+        Returns a user/password touple for the given realm. Note that this function prompts for the
+        password on stdout when the username of the subject is known but no correponding password can be found
+
+        :param realm: realm for the authentication
+        :return: (username, password) touple or None if subject is anonymous or password not found.
+        """
+        if self.anon:
+            return None
+        if self.username is None:
+            if realm in self.hosts_auth:
+                return self.hosts_auth[realm]
+            else:
+                msg = 'No user/password for {}'.format(realm)
+                if self.netrc is not False:
+                    msg = '{} in {}'.format(msg, self.netrc if self.netrc is not True else '$HOME/.netrc')
+                self.logger.error(msg)
+                return None
+        else:
+            if realm in self.hosts_auth \
+                    and self.username == self.hosts_auth[realm][0]:
+                return self.hosts_auth[realm]
+            else:
+                sys.stdout.write("Password for {}@{}: ".format(self.username, realm))
```

Comment:
should there be a fall through to prompt for username if realm not found?  If not, then why is there a password prompt here?